fix: onlyShowFirstAvailableSlot is enabled, but could still bookable via API or by changing the params

RajdeepDs opened the pull request:
## What does this PR do?

This PR fixes an issue where `onlyShowFirstAvailableSlot` was enabled, but could still bookable via the API or by modifying URL params.

- Fixes #16030
- Fixes [CAL-4098](https://linear.app/calcom/issue/CAL-4098/only-show-first-slot-of-day-available-other-slots-still-bookable)

## Updates since last revision

Added server-side validation to prevent bypassing the `onlyShowFirstAvailableSlot` restriction via API or URL manipulation:

- **New validation function** (`validateOnlyShowFirstAvailableSlot.ts`): Validates that the requested booking time is the first available slot of the day when `onlyShowFirstAvailableSlot` is enabled
- **Added `onlyShowFirstAvailableSlot` to DB select** (`getEventTypesFromDB.ts`): Makes the field available in the booking creation flow
- **Integrated validation into booking service** (`RegularBookingService.ts`): Calls validation after `ensureAvailableUsers` populates availability data
- **New error code** (`errorCodes.ts`): Added `BookingNotFirstAvailableSlot` for the validation error

## Visual Demo (For contributors especially)
https://www.loom.com/share/4325894859664c3084edf6912d88ad80?sid=724984fd-f2e8-48fb-8a31-7208f7f4a253

## Human Review Checklist

- [ ] Verify the validation logic correctly identifies the first available slot by comparing with `oooExcludedDateRanges`
- [ ] Check timezone handling edge cases (validation uses event timezone with UTC fallback)
- [ ] Confirm the validation only runs when `skipAvailabilityCheck` is false (intentional for admin/system use)
- [ ] Consider if dedicated unit tests should be added for `validateOnlyShowFirstAvailableSlot`

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] N/A I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

<!-- Link to Devin run: https://app.devin.ai/sessions/60f4836bb7bf42ee86e25140a7320c92 -->
<!-- Requested by: unknown () -->

RajdeepDs committed:
fix: onlyShowFirstAvailableSlot is still bookable via API or by changing the params

- enforced onlyShowFirstAvailableSlot restriction at API level
- prevents bookings for non-first slots, even via API or URL param changes
- added tests to cover onlyShowFirstAvailableSlot behavior

RajdeepDs committed:
fix: enhance onlyShowFirstAvailableSlot functionality in booking logic

- added optional onlyShowFirstAvailableSlot property to BookerEvent type
- refactored unavailable time slot calculation to improve performance and clarity
- updated isAvailableHandler to consider adjacent days when determining the first available slot

TusharBhatt1 commented on packages/features/bookings/Booker/Booker.tsx:
```diff
@@ -191,15 +194,18 @@ const BookerComponent = ({
+  const unavailableTimeSlots = !isQuickAvailabilityCheckFeatureEnabled
+    ? []
+    : allSelectedTimeslots.filter((slot) => {
         return !isTimeSlotAvailable({
```
`isQuickAvailabilityCheckFeatureEnabled ? :` is ideal and more readable

TusharBhatt1 reviewed (commented):
In this case , the message could be more specific and accurate than : 
![image](https://github.com/user-attachments/assets/aa3a9641-3b50-4045-bbd3-8f2c6cdbe254)

@CarinaWolli can you confirm

RajdeepDs committed:
refactor: simplify unavailableTimeSlots logic for readability

kart1ka commented:
Hey @RajdeepDs, Could you please resolve merge conflicts?

RajdeepDs commented:
> Hey @RajdeepDs, Could you please resolve merge conflicts?

Yes, I'll resolve the merge conflicts.

RajdeepDs committed:
Merge branch 'main' into fix-only-show-first-available-slot

kart1ka committed:
Merge branch 'main' into fix-only-show-first-available-slot

kart1ka reviewed (commented):
left a comment.

RajdeepDs commented on packages/trpc/server/routers/viewer/slots/isAvailable.handler.ts:
```diff
@@ -78,6 +79,62 @@ export const isAvailableHandler = async ({
+      slotsByDay[dateStr].push(slot);
+    });
+
+    // For each day, find the first available slot considering slots from previous and next days
+    const filteredSlots: TIsAvailableOutputSchema["slots"] = [];
+
+    Object.keys(slotsByDay).forEach((date) => {
+      const currentDate = dayjs(date);
+      const prevDate = currentDate.subtract(1, "day").format("YYYY-MM-DD");
+      const nextDate = currentDate.add(1, "day").format("YYYY-MM-DD");
+
+      // Get all slots for the current day and adjacent days
+      const allRelevantSlots = [
+        ...(slotsByDay[prevDate] || []),
+        ...(slotsByDay[date] || []),
+        ...(slotsByDay[nextDate] || []),
+      ];
+
```
This code was added because looking up slots only under the current date in GMT can miss valid first-of-day slots for users in other time zones. By including slots from the previous and next days, we ensure that all possible first-available slots are considered, regardless of how they are grouped by date in GMT

https://github.com/calcom/cal.com/pull/21443#discussion_r2100799416

RajdeepDs committed:
Merge branch 'main' into fix-only-show-first-available-slot

devin-ai-integration[bot] committed:
Merge upstream/main into fix-only-show-first-available-slot

Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
Merge upstream/main into fix-only-show-first-available-slot

Co-Authored-By: unknown <>

devin-ai-integration[bot] committed:
fix: add server-side validation for onlyShowFirstAvailableSlot to prevent API bypass

- Add validateOnlyShowFirstAvailableSlot function to validate that the requested
  booking time is the first available slot when onlyShowFirstAvailableSlot is enabled
- Add onlyShowFirstAvailableSlot to getEventTypesFromDBSelect to make the field
  available in the booking creation flow
- Integrate validation into RegularBookingService after ensureAvailableUsers
- Add BookingNotFirstAvailableSlot error code for the validation error

This prevents users from bypassing the onlyShowFirstAvailableSlot restriction
by manipulating URL parameters or calling the API directly.

Co-Authored-By: unknown <>

romitg2 committed:
Merge branch 'main' into fix-only-show-first-available-slot

romitg2 committed:
adds translation key

romitg2 committed:
Merge branch 'main' into fix-only-show-first-available-slot

romitg2 committed:
Merge branch 'main' into fix-only-show-first-available-slot
